src/fstext/determinize-star-inl.h

```diff
@@ -468,99 +469,232 @@ template<class Arc> class DeterminizerStar {
+    std::map<InputStateId, EpsilonClosureInfo> cur_subset;
+
+    typedef typename std::map<InputStateId, EpsilonClosureInfo>::iterator MapIter;
     {
       MapIter iter = cur_subset.end();
-      for (size_t i = 0; i < input_subset.size(); i++) {
-        std::pair<const InputStateId, Element> pr(input_subset[i].state, input_subset[i]);
+      size_t size = input_subset.size();
+      for (size_t i = 0; i < size; i++) {
+        // comment out this because we will process the initial states separately
+        // queue.push(IdAndDepth(input_subset[i].state, 0));
+
+        // the weight has not been processed yet,
+        // so put all of them in the "weight_to_process"
+        EpsilonClosureInfo info(input_subset[i],
+                                input_subset[i].weight,
+                                false,
+                                0);
+        info.element.weight = Weight::Zero(); // clear the weight
+
+        std::pair<const InputStateId, EpsilonClosureInfo>
+                                 pr(input_subset[i].state, info);
         iter = cur_subset.insert(iter, pr);
-        // By providing iterator where we inserted last one, we make insertion more efficient since
+        // By providing iterator where we inserted last one,
+        // we make insertion more efficient since
         // input subset was already in sorted order.
       }
     }
     // find whether input fst is known to be sorted in input label.
-    bool sorted = ((ifst_->Properties(kILabelSorted, false) & kILabelSorted) != 0);
+    bool sorted =
+            ((ifst_->Properties(kILabelSorted, false) & kILabelSorted) != 0);
+
+    size_t size = input_subset.size();
+    for (size_t i = 0; i < size; i++) {
+      ExpandOneElement(input_subset[i],
+                       sorted,
+                       input_subset[i].weight,
+                       &cur_subset,
+                       0);
+    }
 
-    vector<Element> queue(input_subset);  // queue of things to be processed.
-    bool replaced_elems = false; // relates to an optimization, see below.
     int counter = 0; // relates to max-states option, used for test.
     while (queue.size() != 0) {
-      Element elem = queue.back();
-      queue.pop_back();
-      // The next if-statement is a kind of optimization.  It's to prevent us
-      // unnecessarily repeating the processing of a state.  "cur_subset" always
-      // contains only one Element with a particular state.  The issue is that
-      // whenever we modify the Element corresponding to that state in "cur_subset",
-      // both the new (optimal) and old (less-optimal) Element will still be in
-      // "queue".  The next if-statement stops us from wasting compute by
-      // processing the old Element.
-
-      if (replaced_elems && cur_subset[elem.state] != elem)
-        continue;
+      InputStateId id = queue.top().id;
+      int this_depth = queue.top().depth;
+      EpsilonClosureInfo &info = cur_subset[id];
```

Comment:
Actually the code will change it later 

line 680 "elem.weight = Plus(elem.weight, unprocessed_weight);"

Should I not use the reference here, but use the original "cur_subset[id].element.weight = ..." instead?